src/Microsoft.OData.CodeGen/Common/MetadataReader.cs

```diff
@@ -79,32 +95,80 @@ public static Version GetMetadataVersion(ServiceConfiguration serviceConfigurati
-                    {
-                        reader.Read();
-                    }
-
-                    if (reader.EOF)
+                    using (var writer = XmlWriter.Create(workFile))
                     {
-                        throw new InvalidOperationException("The metadata is an empty file");
+                        while (reader.NodeType != XmlNodeType.Element)
+                        {
+                            reader.Read();
+                        }
+
+                        if (reader.EOF)
+                        {
+                            throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, "The metadata is an empty file"));
+                        }
+
+                        Constants.SupportedEdmxNamespaces.TryGetValue(reader.NamespaceURI, out edmxVersion);
+                        writer.WriteNode(reader, false);
                     }
-
-                    Constants.SupportedEdmxNamespaces.TryGetValue(reader.NamespaceURI, out var edmxVersion);
-                    return edmxVersion;
                 }
+
+                return workFile;
             }
             catch (WebException e)
             {
-                throw new InvalidOperationException(string.Format("The metadata cannot be accessed"), e);
+                throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, "Cannot access {0}", serviceConfiguration.Endpoint), e);
+
             }
             finally
             { 
                 metadataStream?.Dispose();
             }
         }
+
+        /// <summary>
+        /// Ensures that the Uri conforms to the expected "$metadata" format
+        /// </summary>
+        /// <param name="uri">Uri to clean</param>
+        /// <returns>Cleaned uri</returns>
+        public static Uri CleanMetadataUri(this Uri uri)
+        {
+            if (uri.Scheme == "http" || uri.Scheme == "https")
+            {
+                UriBuilder uriBuilder;
+
+                /// Evaluates to true if Query and Fragment properties are present in the Uri 
+                bool preserveQueryAndFragment = true;
+
+                if (uri.Segments.Last().Equals("$metadata", StringComparison.InvariantCultureIgnoreCase) | uri.Segments.Last().Equals("$metadata/", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    preserveQueryAndFragment = !uri.AbsolutePath.EndsWith("/", StringComparison.Ordinal);
+                    Uri absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/')).Uri;
+                    uriBuilder = new UriBuilder(absolutePathUri);
+                }
+                else
+                {
+                    var absolutePathUri = new UriBuilder(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath.TrimEnd('/') + "/").Uri;
+                    uriBuilder = new UriBuilder(new Uri(absolutePathUri, "$metadata"));
```

Comment:
```suggestion
                    uriBuilder = new UriBuilder(new Uri(new Uri(uri.Scheme, uri.Host, uri.Port, uri.AbsolutePath), "$metadata"));
```
**Note:** Whether the absolute path ends with `/` or not, the `Uri(Uri, string)` constructor will do the expected thing.

Reply:
`Uri(uri, string)` replaces the last segment if it doesn't end with `/` . For example 
`var baseUri = new Uri("http://localhost/odata");`
`var uri1 = new Uri(baseUri, "$metadata");`
will result to `http://localhost/$metadata`
That is why `/` is added before we add `"$metadata"`